packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/textFeatures.ts

```diff
@@ -0,0 +1,152 @@
+import { createRange, getTagOfNode } from 'roosterjs-editor-dom';
+import { setIndentation } from 'roosterjs-editor-api';
+import {
+    BuildInEditFeature,
+    IEditor,
+    Indentation,
+    TextFeatureSettings,
+    Keys,
+    PluginKeyboardEvent,
+    SelectionRangeTypes,
+    ContentPosition,
+    PositionType,
+    ExperimentalFeatures,
+    NodePosition,
+} from 'roosterjs-editor-types';
+
+/**
+ * Requires @see ExperimentalFeatures.TabKeyTextFeatures to be enabled
+ * Provides additional functionality when press Tab:
+ *      If Whole Paragraph selected, indent paragraph,
+ *      If range is collapsed, add tab spaces
+ *      If range is not collapsed but not all the paragraph is selected, replace selection with Tab spaces
+ */
+const IndentWhenTabText: BuildInEditFeature<PluginKeyboardEvent> = {
+    keys: [Keys.TAB],
+    shouldHandleEvent: (event, editor) =>
+        editor.isFeatureEnabled(ExperimentalFeatures.TabKeyTextFeatures) &&
+        !event.rawEvent.shiftKey &&
+        !editor.getElementAtCursor('LI,TABLE', null /*startFrom*/, event),
+    handleEvent: (event, editor) => {
+        const selection = editor.getSelectionRangeEx();
+        if (selection.type == SelectionRangeTypes.Normal) {
+            editor.addUndoSnapshot(() => {
+                if (selection.areAllCollapsed) {
+                    insertTab(editor, event);
+                } else {
+                    if (isWholeParagraphSelected(editor)) {
+                        setIndentation(editor, Indentation.Increase);
+                    } else {
+                        const { ranges } = selection;
+                        const range = ranges[0];
+                        const tempRange = createRange(range.startContainer, range.startOffset);
+
+                        ranges.forEach(range => range.deleteContents());
+                        editor.select(tempRange);
+                        insertTab(editor, event);
+                    }
+                }
+            });
+
+            event.rawEvent.preventDefault();
+        }
+    },
+};
+
+/**
+ * @internal
+ */
+export const TextFeatures: Record<
+    keyof TextFeatureSettings,
+    BuildInEditFeature<PluginKeyboardEvent>
+> = {
+    indentWhenTabText: IndentWhenTabText,
+};
+
+function isWholeParagraphSelected(editor: IEditor): boolean {
+    const regions = editor.getSelectedRegions();
+    let endPosition: NodePosition = null;
+    let startPosition: NodePosition = null;
+
+    regions.forEach(r => {
+        endPosition = r.fullSelectionEnd;
+        startPosition = startPosition || r.fullSelectionStart;
+    });
```

Comment:
To get first and last position, just simply get them from the selection range. e.g.

```ts
const range = editor.getSelectionRange() // or use getSelectionRangeEx()
const start = Position.getStart(range);
const end=  Position.getEnd(range);
```